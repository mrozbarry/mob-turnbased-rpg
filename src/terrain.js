function* generator(type, noise2dFunction, board) {
  let state = [];
  for(let x = 0; x < board.width; x++) {
    for(let y = 0; y < board.height; y++) {
      if (noise2dFunction(x, y) < 0.8) continue;
      state.push({ type, x, y });
    }
  }

  while (true) {
    yield state;
  }
}

export const make = (type, noise2dFunction, board) => {
  const terrainGenerator = generator(type, noise2dFunction, board);
  terrainGenerator.next();

  return terrainGenerator;
}
