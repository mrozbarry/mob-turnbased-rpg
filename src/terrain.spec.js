import * as Terrain from './terrain.js';
import SimplexNoise from 'simplex-noise';

describe('terrain', () => {
  it('can place rocks in the world', () => {
    const rocks = Terrain.make(
      '*',
      ((x, y) => 1),
      { width: 1, height: 1 },
    );

    const state = rocks.next().value;

    expect(state).toHaveLength(1);
    state.forEach(rock => {
      expect(Object.keys(rock)).toEqual(['type', 'x', 'y']);
      expect(rock.type).toBe('*');
      expect(rock.x).toBe(0);
      expect(rock.y).toBe(0);
    });
  });

  it('can use the noise generator to create stuff', () => {
    const noise = new SimplexNoise('test');
    const rocks = Terrain.make(
      '*',
      ((x,y) => noise.noise2D(x,y)),
      { width: 10, height: 10},
    );

    const state = rocks.next().value;
    console.log(state);

    expect(state).toMatchSnapshot();

  });
});

// https://github.com/jwagner/simplex-noise.js
