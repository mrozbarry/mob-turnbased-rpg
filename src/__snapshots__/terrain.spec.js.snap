// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`terrain can use the noise generator to create stuff 1`] = `
Array [
  Object {
    "type": "*",
    "x": 5,
    "y": 1,
  },
  Object {
    "type": "*",
    "x": 5,
    "y": 4,
  },
  Object {
    "type": "*",
    "x": 7,
    "y": 3,
  },
]
`;
